lib/jormungandr/src/Cardano/Wallet/Jormungandr/Binary.hs

```diff
@@ -537,42 +468,128 @@ getLegacyTransaction tid = do
     let inps = mempty
     pure $ Tx tid inps outs
 
-putSignedTx :: [(TxIn, Coin)] -> [TxOut] -> [TxWitness] -> Put
-putSignedTx inputs outputs witnesses = do
-    putTx inputs outputs
-    unless (length inputs == length witnesses) $
-        fail "number of witnesses must equal number of inputs"
-    mapM_ putWitness witnesses
-  where
-    -- Assumes the `TxWitness` has been faithfully constructed
-    putWitness :: TxWitness -> Put
-    putWitness (TxWitness bytes) = putByteString bytes
-
-putTx :: [(TxIn, Coin)] -> [TxOut] -> Put
-putTx inputs outputs = do
-    unless (length inputs <= fromIntegral (maxBound :: Word8)) $
-        fail $
-            "number of inputs cannot be greater than " ++
-            show maxNumberOfInputs
-    unless (length outputs <= fromIntegral (maxBound :: Word8)) $
-        fail $
-            "number of outputs cannot be greater than " ++
-            show maxNumberOfOutputs
-    putWord8 $ toEnum $ length inputs
-    putWord8 $ toEnum $ length outputs
-    mapM_ putInput inputs
-    mapM_ putOutput outputs
+data MkFragment
+    = MkFragmentSimpleTransaction
+        TxWitnessTag
+    | MkFragmentStakeDelegation
+        TxWitnessTag
+        StakeDelegationType
+        ChimericAccount
+        (XPrv, Passphrase "encryption")
+
+putFragment
+    :: Hash "Genesis"
+    -> [((TxIn, Coin), (XPrv, Passphrase "encryption"))]
+    -> [TxOut]
+    -> MkFragment
+    -> PutM (Proxy "Fragment")
+putFragment (Hash block0Hash) inputs outputs = \case
+    -- SIMPLE-TRANSACTION = TRANSACTION
+    MkFragmentSimpleTransaction witTag ->
+        withHeader FragmentTransaction $ do
+            putTransaction (mempty :: Put) witTag
+
+    -- STAKE-DELEGATION = DLG-CERT TRANSACTION STAKE-AUTH
+    MkFragmentStakeDelegation witTag dlgTag account credentials ->
+        withHeader FragmentDelegation $ do
+            putPayload  <- returnPut $ putDelegationCertificate account dlgTag
+            putAuthData <- returnPut $ putTransaction putPayload witTag
+            putStakeAuthentication (putPayload >> putAuthData) credentials
```

Comment:
Just for the record, what I liked about #1021 was that [this structure](https://github.com/input-output-hk/cardano-wallet/pull/1059/files#diff-8e20179d01c474d1713ed389da590cd8R590-R593) was strictly/automatically enforced such that we wouldn't have to pass in `(putPayload >> putAuthData)` here (for each case that requires it).

But it works :+1: